fractions/microprofile/microprofile-jwt/src/main/java/org/wildfly/swarm/microprofile/jwtauth/deployment/auth/JWTAuthMechanism.java

```diff
@@ -64,54 +65,65 @@ public JWTAuthMechanism(JWTAuthContextInfo authContextInfo) {
+                JWTCredential credential = new JWTCredential(jwtToken, authContextInfo);
+                // Install the JWT principal as the caller
+                Account account = identityManager.verify(credential.getName(), credential);
+                if (account != null) {
+                    JsonWebToken jwtPrincipal = (JsonWebToken) account.getPrincipal();
+                    securityContext.authenticationComplete(account, "MP-JWT", false);
+                    // Workaround authenticated JWTPrincipal not being installed as user principal
+                    // https://issues.jboss.org/browse/WFLY-9212
+                    org.jboss.security.SecurityContext jbSC = SecurityContextAssociation.getSecurityContext();
+                    Subject subject = jbSC.getUtil().getSubject();
+                    jbSC.getUtil().createSubjectInfo(jwtPrincipal, jwtToken, subject);
+                    RoleGroup roles = extract(subject);
+                    jbSC.getUtil().setRoles(roles);
+                    UndertowLogger.SECURITY_LOGGER.debugf("Authenticated caller(%s) for path(%s) with roles: %s",
+                            credential.getName(), exchange.getRequestPath(), account.getRoles());
+                    return AuthenticationMechanismOutcome.AUTHENTICATED;
+                } else {
+                    UndertowLogger.SECURITY_LOGGER.info("Failed to authenticate JWT bearer token");
+                    return AuthenticationMechanismOutcome.NOT_AUTHENTICATED;
                 }
+            } catch (Exception e) {
+                UndertowLogger.SECURITY_LOGGER.infof(e, "Failed to validate JWT bearer token");
+                return AuthenticationMechanismOutcome.NOT_AUTHENTICATED;
             }
         }
 
         // No suitable header has been found in this request,
         return AuthenticationMechanismOutcome.NOT_ATTEMPTED;
     }
 
+    private String getJwtToken(HttpServerExchange exchange) {
+        String bearerToken = null;
+        if (AUTHORIZATION.toString().equals(authContextInfo.getTokenHeader())) {
+            String authScheme = exchange.getRequestHeaders().getFirst(authContextInfo.getTokenHeader());
+            if (authScheme != null && authScheme.toLowerCase(Locale.ENGLISH).startsWith("bearer ")) {
+                bearerToken = authScheme.substring(7);
+            }
+        } else if (COOKIE.toString().equals(authContextInfo.getTokenHeader())
+            && authContextInfo.getTokenCookie() != null) {
+            Cookie cookie = exchange.getRequestCookies().get(authContextInfo.getTokenCookie());
+            if (cookie != null) {
+                bearerToken = cookie.getValue();
+            }
+        } else {
+            bearerToken = exchange.getRequestHeaders().getFirst(authContextInfo.getTokenHeader());
+        }
+        if (bearerToken != null && UndertowLogger.SECURITY_LOGGER.isTraceEnabled()) {
+            UndertowLogger.SECURITY_LOGGER.tracef("Bearer token: %s", bearerToken);
```

Comment:
This looks suspicious. I think we should never ever log secrets, not even on trace level.

Reply:
Right, this is optional as far as I can see, i.e, one needs to enable the security logger. I'm assuming this is why UndetowLogger.SECURITY_LOGGER exists. One possible reason is that a log can be used to confirm that a given token has actually been used for some proof

Reply:
Ah, I see.